src/analysis/processing/qgsalgorithmzonalstatisticsfeaturebased.cpp

```diff
@@ -0,0 +1,189 @@
+  QgsZonalStatistics::Min,
+  QgsZonalStatistics::Max,
+  QgsZonalStatistics::Range,
+  QgsZonalStatistics::Minority,
+  QgsZonalStatistics::Majority,
+  QgsZonalStatistics::Variety,
+  QgsZonalStatistics::Variance,
+};
+
+QString QgsZonalStatisticsFeatureBasedAlgorithm::name() const
+{
+  return QStringLiteral( "zonalstatisticsfb" );
+}
+
+QString QgsZonalStatisticsFeatureBasedAlgorithm::displayName() const
+{
+  return QObject::tr( "Zonal statistics (feature based)" );
+}
+
+QStringList QgsZonalStatisticsFeatureBasedAlgorithm::tags() const
+{
+  return QObject::tr( "stats,statistics,zones,layer,sum,maximum,minimum,mean,count,standard,deviation,"
+                      "median,range,majority,minority,variety,variance,summary,raster" ).split( ',' );
+}
+
+QString QgsZonalStatisticsFeatureBasedAlgorithm::group() const
+{
+  return QObject::tr( "Raster analysis" );
+}
+
+QString QgsZonalStatisticsFeatureBasedAlgorithm::groupId() const
+{
+  return QStringLiteral( "rasteranalysis" );
+}
+
+QString QgsZonalStatisticsFeatureBasedAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm calculates statistics of a raster layer for each feature "
+                      "of an overlapping polygon vector layer." );
+}
+
+QList<int> QgsZonalStatisticsFeatureBasedAlgorithm::inputLayerTypes() const
+{
+  return QList<int>() << QgsProcessing::TypeVectorPolygon;
+}
+
+QgsZonalStatisticsFeatureBasedAlgorithm *QgsZonalStatisticsFeatureBasedAlgorithm::createInstance() const
+{
+  return new QgsZonalStatisticsFeatureBasedAlgorithm();
+}
+
+void QgsZonalStatisticsFeatureBasedAlgorithm::initParameters( const QVariantMap &configuration )
+{
+  Q_UNUSED( configuration )
+  QStringList statChoices;
+  statChoices.reserve( STATS.size() );
+  for ( QgsZonalStatistics::Statistic stat : STATS )
+  {
+    statChoices << QgsZonalStatistics::displayName( stat );
+  }
+
+  addParameter( new QgsProcessingParameterRasterLayer( QStringLiteral( "INPUT_RASTER" ), QObject::tr( "Raster layer" ) ) );
+  addParameter( new QgsProcessingParameterBand( QStringLiteral( "RASTER_BAND" ),
+                QObject::tr( "Raster band" ), 1, QStringLiteral( "INPUT_RASTER" ) ) );
+
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "COLUMN_PREFIX" ), QObject::tr( "Output column prefix" ), QStringLiteral( "_" ) ) );
+
+  addParameter( new QgsProcessingParameterEnum( QStringLiteral( "STATISTICS" ), QObject::tr( "Statistics to calculate" ),
+                statChoices, true, QVariantList() << 0 << 1 << 2 ) );
+}
+
+QString QgsZonalStatisticsFeatureBasedAlgorithm::outputName() const
+{
+  return QObject::tr( "Zonal Statistics" );
+}
+
+QgsFields QgsZonalStatisticsFeatureBasedAlgorithm::outputFields( const QgsFields &inputFields ) const
+{
+  Q_UNUSED( inputFields )
+  return mOutputFields;
+}
+
+QgsProcessingFeatureSource::Flag QgsZonalStatisticsFeatureBasedAlgorithm::sourceFlags() const
+{
+  return QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks;
```

Comment:
Better remove this flag -- the algorithm uses geos intersections so it needs valid inputs.